### PR TITLE
SALTO-1715: added deployChange and addDeploymentAnnotations to adapter-components

### PR DESCRIPTION
**Pull request by alonstern**
Added deployChange and addDeploymentAnnotations to adapter-components

---
_Release Notes_: 
None (Jira not officially released yet)

---
_User Notifications_: 
None

### CODE REVIEW

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  const schemaOrRef = toSchema(SwaggerVersion.V3, operation.requestBody)
+  if (schemaOrRef === undefined) {
+    return []
+  }
+
+  const schema = isReferenceObject(schemaOrRef)
+    ? swagger.parser.$refs.get(schemaOrRef.$ref)
+    : schemaOrRef
+
+  return _(extractProperties(schema, swagger.parser.$refs).allProperties)
```
Comment (netama) [inline packages/adapter-components/src/elements/swagger/deployment/annotations.ts:40]: nit: can we avoid the call to `_`?

**Review by netama — COMMENTED** [review]
cool :) 
added some nits around the config structure, still need to review the swagger implementation parts that I'll try to get to later today (I think most of them at least can also be reviewed by @noamhammer and @ori-moisis ).

a couple of general comments:
1. I think it might be cleaner to have the jira part in a separate PR?
2. I'd like it if we can try to add a bit more in-code documentation in adapter-components. I tried to do it until now (though maybe did not keep it as much in the later additions), would be great to have some short explanation at least in the exported functions that we expect to be reused and are not trivial / require some more context

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts (hunk 1 above)
Comment (alonstern) [reply packages/adapter-components/src/elements/swagger/deployment/annotations.ts:40]: I use `pickBy` which only exists on loadash

File: packages/adapter-components/src/deployment/deployment.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+import { Change, getChangeElement, isInstanceChange, isReferenceExpression, ObjectType, Values, Element } from '@salto-io/adapter-api'
+import { transformValues } from '@salto-io/adapter-utils'
+import _ from 'lodash'
```
[the remark this answers is not in the document]
Comment (alonstern) [reply packages/adapter-components/src/deployment/deployment.ts:18]: 🤷‍♂️

File: packages/adapter-components/src/deployment/deployment.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+export const deployChange = async (
+  change: Change<Element>,
+  client: HTTPWriteClientInterface,
+  endpointDetails?: DeploymentRequests,
+  fieldsToIgnore: string[] = [],
+  additionalUrlVars?: Record<string, string>
+): Promise<ResponseValue | ResponseValue[]> => {
+  if (!isInstanceChange(change)) {
+    throw new Error('Only instance changes are supported')
```
[the remark this answers is not in the document]
Comment (alonstern) [reply packages/adapter-components/src/deployment/deployment.ts:48]: Improved the message but this shouldn't be user facing, it should be caught earlier in the change validator

File: packages/adapter-components/src/deployment/deployment.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+import { Change, getChangeElement, isInstanceChange, isReferenceExpression, ObjectType, Values, Element } from '@salto-io/adapter-api'
+import { transformValues } from '@salto-io/adapter-utils'
+import _ from 'lodash'
+import { getUrlVars, setUrlVarsValues } from '../elements/request_parameters'
+import { HTTPWriteClientInterface } from '../client/http_client'
+import { getDiffInstance } from './diff'
+import { DeploymentRequests } from '../config/request'
+import { ResponseValue } from '../client'
+
+const resolveReferences = async (
```
[the remark this answers is not in the document]
Comment (alonstern) [reply packages/adapter-components/src/deployment/deployment.ts:25]: Moved to `adapter-utils`. It wasn't copied

File: packages/adapter-components/src/config/shared.ts [hunk 5]
```diff
@@ -116,8 +124,8 @@ export const createUserFetchConfigType = (
 )
 
 export const getConfigWithDefault = <
-  T extends TransformationConfig | RequestConfig | undefined,
-  S extends TransformationDefaultConfig | RequestDefaultConfig
+  T extends TransformationConfig | FetchRequestConfig | undefined,
```
[the remark this answers is not in the document]
Comment (alonstern) [reply packages/adapter-components/src/config/shared.ts:127]: Currently this function is never used with the deploy config

File: packages/adapter-components/src/config/request.ts [hunk 6]
```diff
@@ -54,21 +55,31 @@ type RecurseIntoConfig = {
...
   paginationField?: string
 }
 
-export type RequestDefaultConfig = Partial<Omit<RequestConfig, 'url'>>
+export type DeployRequestConfig = BaseRequestConfig & {
+  urlVarsToFields?: Record<string, string>
+  method: 'post' | 'put' | 'delete' | 'patch'
+}
+
+export type DeploymentRequests = Partial<Record<Change['action'], DeployRequestConfig>>
```
[the remark this answers is not in the document]
Comment (alonstern) [reply packages/adapter-components/src/config/request.ts]: `DeploymentRequestByAction`
create the ticket

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts (hunk 1 above)
Comment (ori-moisis) [reply packages/adapter-components/src/elements/swagger/deployment/annotations.ts:40]:
```typescript
Object.keys(_.pickBy(
  extractProperties(schema, swagger.parser.$refs).allProperties,
  isReadOnlyProperty,
))
```

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    .pickBy(val => !('readOnly' in val) || !val.readOnly)
```
Comment (ori-moisis) [inline packages/adapter-components/src/elements/swagger/deployment/annotations.ts:41]: nit: any chance we can extract this and give it a meaningful name, I've been staring at it for 5 minutes and I can't decide if this picks things that are read only or not read only.

File: packages/adapter-components/src/config/request.ts [hunk 8]
```diff
@@ -150,8 +199,8 @@ const findUnresolvedArgs = (url: string, dependsOnArgs: Set<string>): string[] =
 
 export const validateRequestConfig = (
   configPath: string,
-  requestDefaultConfig: RequestDefaultConfig | undefined,
-  requestConfigMap: Record<string, RequestConfig>
+  requestDefaultConfig: FetchRequestDefaultConfig | undefined,
+  requestConfigMap: Record<string, FetchRequestConfig>
```
[the remark this answers is not in the document]
Comment (alonstern) [reply packages/adapter-components/src/config/request.ts:203]: IIUC this only checks the validity of the dependsOn param which does not exist in the deployment params

File: packages/adapter-components/src/deployment/deployment.ts [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+  const resolvedValues = await resolveReferences(
+    _.pickBy(
+      getDiffInstance(change).value,
+      (_value, key) => !fieldsToIgnore.includes(key)
+    ),
+    type
+  )
+
+  const urlVarsValues = Object.fromEntries(
+    getUrlVars(endpoint.url)
```
[the remark this answers is not in the document]
Comment (alonstern) [reply packages/adapter-components/src/deployment/deployment.ts:61]: You mean passing all the values in the instance (plus the urlVarsToFields and additionalUrlVars) and the irrelevant values in the instance just won't be used?

File: packages/adapter-utils/src/diff.ts [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+const removePath = (instance: InstanceElement, path: ElemID): void => {
+  setPath(instance, path, undefined)
+  const parentPath = path.createParentID()
+  if (path.nestingLevel > 1 && _.isEmpty(resolvePath(instance, parentPath))) {
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    removePath(instance, parentPath)
+  }
+}
+
+export const getDiffInstance = (change: Change<InstanceElement>): InstanceElement => {
```
Comment (ori-moisis) [inline packages/adapter-utils/src/diff.ts:32]: this implementation is getting too complicated IMO, I think we will regret not reusing existing operations that contain similar logic (like transform element)

the specific cases for arrays is one example, detecting them in a way that is completely decoupled from the other places in the code that do similar things I think will make this code hard to maintain (it is already quite hard to follow IMO)
but generally building an instance from another instance using `walkOnElement` and not `transformElement` as we usually do I think will lead us to find many more cases that should be handled (and are handled in `transformElement` but are missing from the code below).

will comment on specific bugs below, but I think generally this approach is dangerous and we should try to avoid having to rely on this if we can

File: packages/adapter-utils/src/diff.ts [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    walkOnElement({
+      element: change.data.before,
+      func: ({ value, path }) => {
+        const isValueInArray = isNumber(path.name)
+        if (
+          _.isPlainObject(value)
+          // We don't want to remove values from arrays to not create arrays with "holes" in them
+          || isValueInArray
+        ) {
+          return WALK_NEXT_STEP.RECURSE
```
Comment (ori-moisis) [inline packages/adapter-utils/src/diff.ts:47]: I don't understand this.
the comment says we don't want to remove things from arrays, but then we recurse into the array, so, once we are inside the array, we can remove something and then `removePath` will remove things from the array.
if the intent is to not remove things from arrays, this code doesn't seem to achieve it...

**Review by ori-moisis — COMMENTED** [review]
would approve if it wasn't for the stuff in `getDiffInstance`
maybe we should talk about that f2f again, it seems to be going in a more complicated direction

File: packages/adapter-components/src/deployment/deployment.ts [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    (_value, key) => !fieldsToIgnore.includes(key)
+  )
+
+  const urlVarsValues = {
+    ...instance.value,
+    ..._.mapValues(endpoint.urlVarsToFields ?? {}, fieldName => instance.value[fieldName]),
+    ...(additionalUrlVars ?? {}),
+  }
+  const url = setUrlVarsValues(endpoint.url, urlVarsValues)
+  const response = await client[endpoint.method]({ url, data: valuesToDeploy })
```
[the remark this answers is not in the document]
Comment (alonstern) [reply packages/adapter-components/src/deployment/deployment.ts:58]: the delete function will ignore it since it does not expect `data` in the object param (so it won't be sent to the service)

File: packages/adapter-components/src/deployment/deployment.ts [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+  fieldsToIgnore: string[] = [],
+  additionalUrlVars?: Record<string, string>
+): Promise<ResponseValue | ResponseValue[]> => {
+  const instance = getChangeElement(change)
+  const endpoint = endpointDetails?.[change.action]
+  if (endpoint === undefined) {
+    throw new Error(`No endpoint of type ${change.action} for ${instance.elemID.typeName}`)
+  }
+  const valuesToDeploy = _.pickBy(
+    getDiffInstance(change).value,
```
[the remark this answers is not in the document]
Comment (alonstern) [reply packages/adapter-components/src/deployment/deployment.ts:48]: I probably missunderstood, fixed it now

File: packages/adapter-utils/src/diff.ts (hunk 10 above)
Comment (alonstern) [reply packages/adapter-utils/src/diff.ts:32]: Removed the special behavior for arrays (not needed anymore because this is not used in deploy)

I'm having a bit of trouble using `transformElement` instead of `walkOnElement` because I'm not sure how I can delete objects that became empty because they all had the same values without deleting empty objects that really became empty in the after part of the change

File: packages/adapter-utils/src/diff.ts (hunk 11 above)
Comment (alonstern) [reply packages/adapter-utils/src/diff.ts:47]: removed

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts (hunk 7 above)
Comment (alonstern) [reply packages/adapter-components/src/elements/swagger/deployment/annotations.ts:42]: I extracted it to `editableFieldNames`, is it still not clear?

File: packages/adapter-components/src/deployment/deployment.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/adapter-components/src/deployment/deployment.ts:76]: I think both options are potentially bad, that is the problem with making things infrastructure so early...
doesn't matter what you choose to do here, it would potentially break the next adapter

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts (hunk 7 above)
Comment (ori-moisis) [reply packages/adapter-components/src/elements/swagger/deployment/annotations.ts:42]: so it is "not read only"?

**Review by ori-moisis — APPROVED** [review]
I would take the whole diff instance thing out of this, I think it is adding noise and is not actually needed (and still has more bugs IIUC), but up to you
the rest of the comments are nits, so you can pick what to do with them

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts (hunk 7 above)
Comment (alonstern) [reply packages/adapter-components/src/elements/swagger/deployment/annotations.ts:42]: yes

File: packages/adapter-components/src/config/request.ts [hunk 14]
```diff
@@ -117,30 +140,59 @@ export const createRequestConfigs = (
...
+      ...sharedEndpointFields,
+      method: {
         refType: BuiltinTypes.STRING,
         annotations: {
-          [CORE_ANNOTATIONS.REQUIRED]: true,
+          [CORE_ANNOTATIONS.RESTRICTION]: createRestriction({ values: ['get', 'post', 'put', 'delete', 'patch'] }),
         },
       },
-      ...sharedEndpointFields,
+      urlVarsToFields: {
```
[the remark this answers is not in the document]
Comment (alonstern) [reply packages/adapter-components/src/config/request.ts:168]: `queryParams` are they http url params, meaning { a: 'b' }  => `http://<url>?a=b`
`urlVarsToFields` is for filling the template values in the url, e.g. if the url is `http://url/{someId}` 
and `urlVarsToFields` is `{ someId: 'id' }`, we will take fill the url with the value of the `id` field in the instance

File: packages/adapter-components/src/elements/swagger/type_elements/swagger_parser.ts [hunk 15]
```diff
@@ -109,44 +110,62 @@ export type SchemasAndRefs = {
   refs: SwaggerRefs
 }
 
-export const getParsedDefs = async (swaggerPath: string):
-  Promise<SchemasAndRefs> => {
-  const parser = new SwaggerParser()
-  const parsedSwagger: OpenAPI.Document = await parser.bundle(swaggerPath)
+type V2Def = { schema: OpenAPIV2.Schema }
```
Comment (ori-moisis) [inline packages/adapter-components/src/elements/swagger/type_elements/swagger_parser.ts:113]: no idea what is going on here (don't fully understand the old code)
@netama wants to review this so I'll dismiss my approval and she will approve after reviewing this part

File: packages/adapter-components/src/elements/swagger/type_elements/swagger_parser.ts [hunk 16]
```diff
@@ -109,44 +110,62 @@ export type SchemasAndRefs = {
...
+  def?.schema
 
-  const toSchemaV3 = (
-    def?: OpenAPIV3.OperationObject
-  ): undefined | OpenAPIV3.SchemaObject | OpenAPIV3.ReferenceObject => {
-    const response = def?.responses?.[200]
-    if (isReferenceObject(response)) {
-      return response
-    }
+type V3Def = OpenAPIV3.ReferenceObject | OpenAPIV3.ResponseObject |OpenAPIV3.RequestBodyObject
```
Comment (netama) [inline packages/adapter-components/src/elements/swagger/type_elements/swagger_parser.ts:118]: nit: missing space after last `|`

File: packages/adapter-components/src/elements/swagger/type_elements/swagger_parser.ts [hunk 17]
```diff
@@ -109,44 +110,62 @@ export type SchemasAndRefs = {
...
+  return undefined
+}
+
+export enum SwaggerVersion {
+  V2,
+  V3
+}
 
-  if (!(isV2(parsedSwagger) || isV3(parsedSwagger))) {
+export const getSwaggerVersion = (swagger: LoadedSwagger): SwaggerVersion => {
```
Comment (netama) [inline packages/adapter-components/src/elements/swagger/type_elements/swagger_parser.ts:141]: looks like this is only used in this file, do we need to export?

**Review by netama — APPROVED** [review]
LGTM! (focused on the swagger stuff following @ori-moisis 's previous reivew) 
added a few nits, and one comment that I think might be a bug but maybe I'm missing something. once these are resolved should be good to merge 👍

File: packages/adapter-components/src/config/request.ts [hunk 18]
```diff
@@ -54,21 +55,33 @@ type RecurseIntoConfig = {
...
 
-export type RequestDefaultConfig = Partial<Omit<RequestConfig, 'url'>>
+export type UrlParams = Record<string, string>
+
+export type DeployRequestConfig = BaseRequestConfig & {
+  urlParamsToFields?: UrlParams
+  method: 'post' | 'put' | 'delete' | 'patch'
+}
+
+export type DeploymentRequestsByAction = Partial<Record<ActionName, DeployRequestConfig>>
```
[the remark this answers is not in the document]
Comment (alonstern) [reply packages/adapter-components/src/config/request.ts:77]: we can discuss it more IRL but as I said before, changing it to create/update/delete will mean we will need another map in our code:
```
const map = {
    add: 'create'
    modify: 'update'
    remove: 'delete'
}
```
which IMO is redundant and very confusing since we end up with two terminologies in our code that means the same thing

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts (hunk 7 above)
Comment (alonstern) [reply packages/adapter-components/src/elements/swagger/deployment/annotations.ts:42]: I want to get the values that do not have a `readOnly = true` property.

`val` is `unknown` so I can't access to `readOnly` before I check that it is in the object
`val.readOnly` does not equal to `true` if `val` does not have a 'readOnly' value i.e., `!('readOnly' in val)`
or if it has a readOnly value, but it is not `true`, i.e., `!val.readOnly`
